### PR TITLE
Stream cache error handling improvements.

### PR DESCRIPTION
**Pull request by jason-bragg**
Error handling of queue cache exceptions was limited to DataNotAvaiable exception, and cache cursor was not being cleaned up correctly.

Queue cache exceptions (of all types) are now handled correctly, and cache cursor's are cleaned up during exceptions.

### CODE REVIEW

**Comment by gabikliot** [comment]
Looks good to me.

**Comment by jason-bragg** [comment]
Assigned to @sergeybykov for merge.

File: src/OrleansRuntime/Streams/PersistentStream/PersistentStreamPullingAgent.cs [hunk 1]
```diff
@@ -555,17 +563,8 @@ private void StartInactiveCursors(StreamConsumerCollection streamData)
...
-                consumerData.LastToken = newToken;
-                consumerData.Cursor = queueCache.GetCacheCursor(consumerData.StreamId.Guid,
-                    consumerData.StreamId.Namespace, newToken.Token);
-            }
-            else
-            {
-                consumerData.LastToken = StreamHandshakeToken.CreateDeliveyToken(batch.SequenceToken); // this is the currently delivered token
-            }
-
+            consumerData.LastToken = StreamHandshakeToken.CreateDeliveyToken(batch.SequenceToken); // this is the currently delivered token
```
Comment (gabikliot) [inline src/OrleansRuntime/Streams/PersistentStream/PersistentStreamPullingAgent.cs]: Don't you want to do that only in the else case? Only if new token is null?

Comment (gabikliot) [inline src/OrleansRuntime/Streams/PersistentStream/PersistentStreamPullingAgent.cs]: Better to move that line also to outside that function, just like you did to the rest of the code.

Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PersistentStream/PersistentStreamPullingAgent.cs]: It does not matter and code is simpler this way.  Think of it like this:
1) We deliver the event.
2) When the delivery is completed, we update the last token.
3) If a correction was returned (newToken != null), then we update the last token from the correction and get a new cursor.
This cleanly separates the delivery functionality for the handling of the response.

**Pull request merged**